src/util/LazyJsonParser.cpp

```diff
@@ -0,0 +1,258 @@
+    }
+    if (isEscaped_) {
+      isEscaped_ = false;
+      continue;
+    }
+    switch (input_[idx]) {
+      case '"':
+        // End of literal.
+        if (std::holds_alternative<BeforeArrayPath>(state_)) {
+          std::get<BeforeArrayPath>(state_).litLength_ =
+              idx - std::get<BeforeArrayPath>(state_).litStart_;
+        }
+        inLiteral_ = false;
+        return;
+      case '\\':
+        isEscaped_ = true;
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseBeforeArrayPath(size_t& idx) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<BeforeArrayPath>(state_));
+  auto& state = std::get<BeforeArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx >= 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Header size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+        state.tryAddKeyToPath(input_);
+        break;
+      case '[':
+        if (state.openBrackets_ == 0) {
+          state.tryAddKeyToPath(input_);
+        }
+        ++state.openBrackets_;
+        if (state.curPath_ == arrayPath_) {
+          // Reached arrayPath.
+          state_ = InArrayPath{.arrayStartIdx_ = idx};
+          ++idx;
+          return;
+        }
+        break;
+      case ']':
+        --state.openBrackets_;
+        if (state.openBrackets_ == 0 && !state.curPath_.empty()) {
+          state.curPath_.pop_back();
+        }
+        break;
+      case '}':
+        if (!state.curPath_.empty()) {
+          state.curPath_.pop_back();
+        }
+        break;
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseInArrayPath(size_t& idx, size_t& materializeEnd) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<InArrayPath>(state_));
+  auto& state = std::get<InArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx - std::max(state.arrayStartIdx_, materializeEnd) >= 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+      case '[':
+        ++state.openBracketsAndBraces_;
+        break;
+      case '}':
+        --state.openBracketsAndBraces_;
+        break;
+      case ']':
+        if (state.openBracketsAndBraces_ == 0) {
+          // End of ArrayPath reached.
+          state_ = AfterArrayPath{.remainingBraces_ = arrayPath_.size(),
+                                  .arrayEndIdx_ = idx};
+          ++idx;
+          return;
+        }
+        --state.openBracketsAndBraces_;
+        break;
+      case ',':
+        if (state.openBracketsAndBraces_ == 0) {
+          materializeEnd = idx;
```

Comment:
As this `materializeEnd` is not read, but only written, it should be a return value of this function
(together with a comment, what this function returns).